design-proposals/klip-33-key-format.md

```diff
@@ -0,0 +1,299 @@
+# KLIP 33 - Key format
+
+**Author**: @big-andy-coates | 
+**Release Target**: 0.13 | 
+**Status**: In Discussion | 
+**Discussion**: _link to the design discussion PR_
+
+**tl;dr:** ksqlDB currently only supports keys compatible with the `KAFKA` format. This limits the
+           data ksqlDB can work with. Extending the set of key formats ksqlDB supports immediately
+           opens up the use of ksqlDB with previously incompatible datasets.
+           
+## Motivation and background
+
+Data stored in Kafka has a key and a value. These can be serialized using different formats, but are
+generally use a common format. ksqlDB supports multiple _value_ [data formats][1], but requires the 
+key data format to be the `KAFKA`.  
+
+This limitation is particularly problematic for tables. ksqlDB is unable to access changelog topics
+in Kafka that have non-`KAFKA` formatted keys. As the key of the Kafka record is the `PRIMARY KEY` 
+of the table, it is essential that the key can be read if the changelog is to be materialised into a 
+table. When changelog topics have non-`KAFKA` key formats, the limitation precludes ksqlDB as a solution.
+
+The limitation is less damaging for streams. However, it is still the case that the user looses the
+ability to access the data in the Kafka record's key. If this data is not duplicated in the record's
+value, which generally seems to be the case, then the data is not accessible at all. If the data is
+required, then the limitation precludes ksqlDB as a solution.
+
+As well as unsupported _input_ key formats, ksqlDB is equally precluded should a solution require the 
+_output_ to have a non-`KAFKA` key format. ksqlDB is already often used as the glue between 
+disparate systems, even though it is limited to changing the _values_ format and structure.
+Supporting other key formats opens this up to also transforming the key into a different format. 
+
+In some cases users are able to work around this limitation. This may involve changing upstream code,
+or introducing pre-processing, or, in the case of Connect, using SMTs to convert the key format. All
+such solutions tend to increase the complexity of the system, and generally hurt performance.
+
+To open ksqlDB up to new problems spaces and to drive adoption, ksqlDB should support other key formats. 
+
+## What is in scope
+
+ * Addition of a new optional `KEY_FORMAT` property in the `WITH` clause, to set the key format.
+ * Addition of a new optional `FORMAT` property in the `WITH` clause, to set both the key & value formats.
+ * Addition of new server configuration to provide defaults for key and value formats.
+ * Removal of requirement for `VALUE_FORMAT` property in the `WITH` clause of CT/CS statements, where
+   the server configuration provides a default.
+ * Support for additional key column data types, where the key format supports it:
+    * `DECIMAL`
+    * `BOOLEAN`
+ * Support of the following key formats:
```

Comment:
I suspect `KAFKA` will be supported as well?

Reply:
It's already supported. So... adding support is not in scope.

Reply:
Well, but `KEY_FORMAT='KAFKA'` should be a valid option thus it seem it should be listed? It's not about adding support for the format, but about the language change to explicitly set the format.